File: javascript/queen-attack/queen-attack.js
export class QueenAttack {
    constructor({
        black: [blackRow, blackColumn] = [0, 3],
        white: [whiteRow, whiteColumn] = [7, 3],
    } = {}) {
        if ([blackRow, blackColumn, whiteRow, whiteColumn].some(n => n < 0 || n > 7)) {
            throw new Error('Queen must be placed on the board');
        }

        if (blackRow === whiteRow && blackColumn === whiteColumn) {
            throw new Error('Queens cannot share the same space')
        }

        this.black = [blackRow, blackColumn]
        this.white = [whiteRow, whiteColumn]
    }

    toString() {
        let board = [...Array(8)].map(() => Array(8).fill('_'));
        board[this.black[0]][this.black[1]] = 'B';
        board[this.white[0]][this.white[1]] = 'W';

        return board.map(row => row.join(' ')).join('\n');

    }

    get canAttack() {
        return this.black[0] === this.white[0] || this.black[1] === this.white[1]
            || this.black[0] + this.black[1] === this.white[0] + this.white[1]
            || Math.abs(this.black[0] - this.black[1]) === Math.abs(this.white[0] - this.white[1]);
    }
}

File: python/queen-attack/queen_attack.py
class Queen:
    def __init__(self, row, column):

        # if the row parameter is negative
        if row < 0:
            raise ValueError("row not positive")

        # if the row parameter is not on the defined board
        if row > 7:
            raise ValueError("row not on board")

        # if the column parameter is negative
        if column < 0:
            raise ValueError("column not positive")

        # if the column parameter is not on the defined board
        if column > 7:
            raise ValueError("column not on board")

        self.row = row
        self.column = column

    def can_attack(self, another_queen):

        # if both the queens are on the same location
        if self.row == another_queen.row and self.column == another_queen.column:
            raise ValueError("Invalid queen position: both queens in the same square")

        if self.row == another_queen.row or self.column == another_queen.column:
            return True

        if self.row - self.column == another_queen.row - another_queen.column:
            return True

        if self.row + self.column == another_queen.row + another_queen.column:
            return True

        return False
